indico/modules/rb_new/client/js/common/bookings/BookingEditForm.jsx

```diff
@@ -268,17 +275,31 @@ class BookingEditForm extends React.Component {
                     )}
                     <Field name="dates"
                            isSingleBooking={recurrence.type === 'single'}
-                           onChange={(newDates) => onBookingPeriodChange(newDates, timeSlot, recurrence)}
+                           onChange={(newDates) => {
+                               this.setState({formChanged: true});
+                               onBookingPeriodChange(newDates, timeSlot, recurrence);
+                           }}
                            disabled={submitSucceeded}
                            isEqual={_.isEqual}
                            render={this.renderDateForm} />
                     {!hideOptions.timeSlot && (
                         <Field name="timeSlot"
-                               onChange={(newTimeSlot) => onBookingPeriodChange(dates, newTimeSlot, recurrence)}
+                               onChange={(newTimeSlot) => {
+                                   this.setState({formChanged: true});
+                                   onBookingPeriodChange(dates, newTimeSlot, recurrence);
+                               }}
                                disabled={submitSucceeded}
                                isEqual={_.isEqual}
                                render={this.renderTimeForm} />
                     )}
+                    {formChanged && showWarning && (
```

Comment:
You can use a FormSpy instead of tracking the state manually (which doesn't reset e.g. when you switch back to the old value). Something like this:

```jsx
{!room.canUserBook && room.canUserPrebook && (
    <FormSpy subscription={{dirtyFields: true}}>
        {({dirtyFields}) => (
            !!_.intersection(dirtyFields, ['recurrence.type', ...]).length && (
                <Message ...>...</Message>
            )
        )}
    </FormSpy>
)}
```

Reply:
I just checked again and dirtyFields is a dict mapping field names to booleans and not just an array as I initially thought. But it never contains false values, so you could just throw in an `Object.keys()` there... Or better, add a util function that checks the relevant keys. Like this you could also use `startsWith('recurrence.')` instead of listing every `recurrence.*` name.

Reply:
I noticed that and added `Object.keys()`. I also noticed that the intersection should be between the `timeSlot` and `dates` and not with anything related to recurrences but when you switch from `single` to anything else the dates are also affected, thus the warning appears.

Reply:
Are you sure the dates will become dirty if you change the recurrence in a way where the start and end date are still the same?

Reply:
I think it applies only when you go from single -> to something else since there is an `endDate` that is added there